cpp/dolfinx/la/SparsityPattern.cpp

```diff
@@ -238,83 +239,94 @@ void SparsityPattern::assemble()
   for (std::int64_t global_i : ghosts1)
     global_to_local.insert({global_i, local_i++});
 
-  // For each ghost row, pack and send (global row, global col) pairs to send to
-  // neighborhood
-  std::vector<std::int64_t> ghost_data;
+  // For each ghost row, pack and send (global row, global col, col_owner)
+  // triplets to send to neighborhood
+  std::map<int, std::vector<std::int64_t>> ghost_data;
```

Comment:
This can be a very expensive data structure. We have a local index for the ghost columns, so no need to use a map. You should be able to use an adjacency list, which can then be used directly with alltoallv. Just need a first loop to compute the row sizes in the adjacency list.